File: sdk/src/units.js
// Amounts.
//
// Every amount on Hikmalayer is an integer number of base units, and the SDK
// represents them as `BigInt` without exception. This is not fussiness: HKM
// has 6 decimals and a 100-billion supply, so the largest legitimate amount
// is 10^17 base units — well past `Number.MAX_SAFE_INTEGER` (2^53−1 ≈
// 9.007×10^15). A `number` cannot hold it.
//
// The failure is quiet and expensive. Signatures cover the exact decimal text
// of an amount, so a value that lost its low digits produces a transaction
// that no longer matches what was signed: the node rejects an honest transfer
// and nothing in the error explains why. So: BigInt in, decimal strings on
// the wire, and never a float in between.

/// Decimals of the native coin.
export const HKM_DECIMALS = 6;

/// Base units in one HKM.
export const UNITS_PER_HKM = 1_000_000n;

/// Coerce to BigInt, rejecting the lossy inputs rather than accepting them.
///
/// A `number` is allowed only when it is a safe integer — the one case where
/// it is provably exact. Anything larger must arrive as a string or BigInt.
export function toBaseUnits(value) {
  if (typeof value === "bigint") return value;
  if (typeof value === "string") {
    const text = value.trim();
    if (!/^-?\d+$/.test(text)) {
      throw new TypeError(
        `Expected a whole number of base units, got ${JSON.stringify(value)}. ` +
          `To convert a decimal amount like "1.5", use parseUnits().`
      );
    }
    return BigInt(text);
  }
  if (typeof value === "number") {
    if (!Number.isInteger(value)) {
      throw new TypeError(`Base units must be a whole number, got ${value}`);
    }
    if (!Number.isSafeInteger(value)) {
      throw new TypeError(
        `${value} exceeds Number.MAX_SAFE_INTEGER and has already lost precision. ` +
          `Pass a BigInt or a decimal string instead.`
      );
    }
    return BigInt(value);
  }
  throw new TypeError(`Cannot interpret ${typeof value} as an amount`);
}

/// Parse a human decimal string into base units: ("1.5", 6) -> 1500000n.
///
/// Rejects more precision than the asset supports rather than rounding it
/// away — a truncated amount is a different amount, and the user should hear
/// about it before signing.
export function parseUnits(input, decimals = HKM_DECIMALS) {
  const text = String(input ?? "").trim();
  if (!text) return 0n;
  if (!/^-?\d*\.?\d*$/.test(text) || text === "." || text === "-") {
    throw new TypeError(`"${input}" is not a decimal number`);
  }
  const negative = text.startsWith("-");
  const body = negative ? text.slice(1) : text;
  const [whole = "", fraction = ""] = body.split(".");
  if (fraction.length > decimals) {
    throw new TypeError(
      `This asset has ${decimals} decimals; "${input}" has ${fraction.length}`
    );
  }
  const units = BigInt(`${whole || "0"}${fraction.padEnd(decimals, "0")}`);
  return negative ? -units : units;
}

/// Format base units for display: (1234567n, 6) -> "1.234567".
export function formatUnits(baseUnits, decimals = HKM_DECIMALS) {
  let value = toBaseUnits(baseUnits);
  if (decimals === 0) return value.toString();
  const negative = value < 0n;
  if (negative) value = -value;
  const scale = 10n ** BigInt(decimals);
  const fraction = (value % scale).toString().padStart(decimals, "0").replace(/0+$/, "");
  const text = fraction ? `${value / scale}.${fraction}` : (value / scale).toString();
  return negative ? `-${text}` : text;
}

/// Serialize an amount for the wire. Always a decimal string: exact by
/// construction, and accepted by every amount field in the node API.
export function encodeAmount(value) {
  return toBaseUnits(value).toString();
}

/// Apply a slippage tolerance (in percent) to an expected output, returning
/// the minimum acceptable amount. This is what a signature commits to, and
/// therefore what stops a trade being sandwiched between signing and
/// execution. Computed in basis points so the arithmetic stays integral.
export function applySlippage(expectedOut, slippagePercent) {
  const expected = toBaseUnits(expectedOut);
  const bps = BigInt(Math.round(Number(slippagePercent) * 100));
  if (bps <= 0n) return expected;
  if (bps >= 10_000n) return 0n;
  return (expected * (10_000n - bps)) / 10_000n;
}

/// Integer square root, matching `isqrt_u128` in the node. Used for the LP
/// shares minted by a pool's first deposit.
export function isqrt(value) {
  const n = toBaseUnits(value);
  if (n < 0n) throw new RangeError("isqrt of a negative number");
  if (n < 2n) return n;
  let x = n;
  let y = (n + 1n) / 2n;
  while (y < x) {
    x = y;
    y = (x + n / x) / 2n;
  }
  return x;
}

File: python-sdk/hikmalayer/units.py
"""
Amount handling for Hikmalayer.

All on-chain amounts are integer BASE UNITS. Python's int is arbitrary
precision, so unlike JavaScript there is no 2^53 cliff — but the same
discipline applies: never use float for an amount, and never send a
float to the node.

    1 HKM = 1_000_000 base units (6 decimals)

Each native token (HTS) declares its own decimal count.
"""

from decimal import Decimal, InvalidOperation
from typing import Union

HKM_DECIMALS = 6
HKM_SCALE = 10 ** HKM_DECIMALS

Amount = Union[int, str, Decimal]


def parse_units(value: Union[str, int, Decimal], decimals: int = HKM_DECIMALS) -> int:
    """
    Parse a human-readable decimal amount into integer base units.

    >>> parse_units("1.5", 6)
    1500000
    >>> parse_units("10", 6)
    10000000
    >>> parse_units("0.000001", 6)
    1

    Raises ValueError on malformed input or excess precision.
    """
    text = str(value).strip()
    if not text:
        return 0

    if text.startswith("-"):
        raise ValueError("Amount must not be negative")

    try:
        dec = Decimal(text)
    except InvalidOperation as exc:
        raise ValueError(f"Amount must be a decimal number: {value!r}") from exc

    sign, digits, exponent = dec.as_tuple()
    if exponent < 0 and -exponent > decimals:
        raise ValueError(f"At most {decimals} decimal places for this asset")

    scaled = dec * (10 ** decimals)
    if scaled != scaled.to_integral_value():
        raise ValueError(f"At most {decimals} decimal places for this asset")

    return int(scaled)


def format_units(base_units: Amount, decimals: int = HKM_DECIMALS) -> str:
    """
    Format integer base units as a human-readable decimal string.

    >>> format_units(1500000, 6)
    '1.5'
    >>> format_units(1000000, 6)
    '1'
    >>> format_units(1, 6)
    '0.000001'
    """
    value = int(base_units)
    if decimals == 0:
        return str(value)

    negative = value < 0
    if negative:
        value = -value

    scale = 10 ** decimals
    whole = value // scale
    fraction = str(value % scale).rjust(decimals, "0").rstrip("0")

    text = f"{whole}.{fraction}" if fraction else str(whole)
    return f"-{text}" if negative else text


def parse_hkm(value: Union[str, int, Decimal]) -> int:
    """Shorthand: parse an HKM amount (6 decimals) into base units."""
    return parse_units(value, HKM_DECIMALS)


def format_hkm(base_units: Amount) -> str:
    """Shorthand: format HKM base units as a decimal string."""
    return format_units(base_units, HKM_DECIMALS)


def to_base_units(value: Amount) -> int:
    """
    Coerce an amount to an integer, refusing anything that has already
    lost precision. Mirrors the JS SDK's toBaseUnits.
    """
    if isinstance(value, bool):
        raise TypeError("A boolean is not an amount")
    if isinstance(value, int):
        return value
    if isinstance(value, Decimal):
        if value != value.to_integral_value():
            raise ValueError(f"{value} is not a whole number of base units")
        return int(value)
    if isinstance(value, str):
        text = value.strip()
        if not text.lstrip("-").isdigit():
            raise ValueError(f"{value!r} is not an integer amount")
        return int(text)
    if isinstance(value, float):
        raise TypeError(
            "A float amount has already lost precision. "
            "Pass an int or a decimal string instead."
        )
    raise TypeError(f"Cannot interpret {type(value).__name__} as an amount")


def apply_slippage(amount: int, tolerance_bps: int) -> int:
    """
    Apply a slippage tolerance in basis points to a quoted amount, producing
    a minimum-out bound. 50 bps = 0.5%.

    >>> apply_slippage(1000000, 50)
    995000
    """
    if not 0 <= tolerance_bps <= 10_000:
        raise ValueError("Tolerance must be between 0 and 10000 basis points")
    return (int(amount) * (10_000 - tolerance_bps)) // 10_000


def isqrt(n: int) -> int:
    """
    Integer square root, matching the chain's implementation exactly.
    Used for predicting LP shares on a first liquidity deposit.
    """
    if n < 0:
        raise ValueError("isqrt of a negative number")
    if n < 2:
        return n
    x = n
    y = (x + 1) // 2
    while y < x:
        x = y
        y = (x + n // x) // 2
    return x
